kernel/smp.c

```diff
@@ -64,62 +103,128 @@ static void wait_for_start_signal(atomic_t *start_flag)
 	}
 }
 
-/* Legacy interfaces for early-version SOF CPU bringup.  To be removed */
-#ifdef CONFIG_SOF
-void z_smp_thread_init(void *arg, struct k_thread *thread)
-{
-	z_dummy_thread_init(thread);
-	wait_for_start_signal(arg);
-}
-void z_smp_thread_swap(void)
-{
-	z_swap_unlocked();
-}
-#endif
-
 static inline FUNC_NORETURN void smp_init_top(void *arg)
 {
 	struct k_thread dummy_thread;
+	struct cpu_start_cb *csc = arg;
 
+	/* Let start_cpu() know that this CPU has powered up. */
 	(void)atomic_set(&ready_flag, 1);
 
-	wait_for_start_signal(arg);
+	/* Wait for the CPU start caller to signal that
+	 * we can start initialization.
+	 */
+	wait_for_start_signal(&cpu_start_flag);
+
+	/* Initialize the dummy thread struct so that
+	 * the scheduler can schedule actual threads to run.
+	 */
 	z_dummy_thread_init(&dummy_thread);
+
 #ifdef CONFIG_SYS_CLOCK_EXISTS
-	smp_timer_init();
+	if ((csc == NULL) || ((csc != NULL) && csc->reinit_timer)) {
```

Comment:
Style: the `csc != NULL` test is needless

Reply:
Removed.